mlir/lib/Transforms/SROA.cpp

```diff
@@ -0,0 +1,194 @@
+//===-- SROA.cpp - Scalar Replacement Of Aggregates -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Transforms/SROA.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Interfaces/MemorySlotInterfaces.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_SROA
+#include "mlir/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+Optional<MemorySlotDestructionInfo>
+mlir::computeDestructionInfo(DestructibleMemorySlot &slot) {
+  assert(isa<DestructibleTypeInterface>(slot.elemType));
+
+  MemorySlotDestructionInfo info;
+
+  SmallVector<MemorySlot> usedSafelyWorklist;
+
+  auto scheduleAsBlockingUse = [&](OpOperand &use) {
+    SmallPtrSet<OpOperand *, 4> &blockingUses =
+        info.userToBlockingUses.getOrInsertDefault(use.getOwner());
+    blockingUses.insert(&use);
+  };
+
+  // Initialize the analysis with the immediate users of the slot.
+  for (OpOperand &use : slot.ptr.getUses()) {
+    if (auto accessor =
+            dyn_cast<DestructibleAccessorOpInterface>(use.getOwner())) {
+      if (accessor.canRewire(slot, info.usedIndices, usedSafelyWorklist)) {
+        info.accessors.push_back(accessor);
+        continue;
+      }
+    }
+
+    // If it cannot be shown that the operation uses the slot safely, maybe it
+    // can be promoted out of using the slot?
+    scheduleAsBlockingUse(use);
+  }
+
+  SmallPtrSet<OpOperand *, 16> dealtWith;
```

Comment:
I would either use something more descriptive, or revert to `seen`. "Dealt with" sounds a bit informal 😁